### ISSUE TITLE
Grpc client - inconsistent Request.IsInProcessRequest()

### ISSUE THREAD

**Issue by ludydoo** [ServiceStack/Issues#694]
#### Configuration

ServiceStack v.5.7.1 (MyGet)
Framework .NET Core 3
Visual Studio 2019 Community
Windows 10

I am calling the `RegisterService` from another service (`TenantService`). 

When a user sends a `SignUpTenant` request, the `TenantService` creates the tenant (initializes the database, ...), and registers the user that signed up by doing `Gateway.SendAsync(registerDto);`

I set up a few integration test classes. I have a base class `IntegrationTestBase`, in which I set up a `[OneTimeSetup]` method.

In this method, I create a `Tenant` by sending a `SignupTenant` dto, so that each integration tests can use this `Tenant` for themselves. Also, in this `[OneTimeSetup]`, I create a new `IntegrationTestsAppHost`. So each integration test class has its own `AppHost`. Each AppHost is running on its own port.

I restricted the RegisterService by checking `if (!req.IsInProcessRequest()){ throw HttpError.MethodNotAllowed("Not allowed"); }`. I want to restrict the usage of the `RegisterService` to only other services (in this case, only `TenantService` is using it). 

When I run a single unit test, this works fine. I can see that the `InProcess` request attribute is successfully set. (sometimes)

But when multiple test are run (either for the same Integration test class, or for all integrationt test ran at the same time), the value of `IsInProcessRequest()` is inconsistent. When I run the test in debug mode, I believe the execution time is slower. So the `IsInProcessRequest()` always returns true.

But when I run the tests without debug mode, I can see that the `IsInProcessRequest()` is almost always false.

I also tried to set the `[assembly: Parallelizable(ParallelScope.None)]` but it did not help.

`IntegrationTestBase.cs`
```
        [OneTimeSetUp]
        public virtual Task OneTimeSetup()
        {
            _context = new IntegrationTestContext();
            return Task.CompletedTask;
        }

        [OneTimeTearDown]
        public virtual void OneTimeTearDown()
        {
            _context?.Dispose();
        }
```

`IntegrationTestContext.cs`
```
        public static int Count;
        private readonly string           _baseUri;
        private readonly ServiceStackHost _appHost;

        public IntegrationTestContext()
        {
            _baseUri = $"http://localhost:{2000 + Count++}/";
            _appHost = new ItAppHost();
            _appHost.Init();
            _appHost.Start(_baseUri);
            GrpcClientFactory.AllowUnencryptedHttp2 = true;
        }

        public void Dispose()
        {
            _appHost?.Dispose();
        }

        public GrpcServiceClient CreateClient()
        {
            var client = new GrpcServiceClient(_baseUri) {UserAgent = "gRPC"};
            return client;
        }
```
`CustomRegisterService.cs`
```
        public async Task<object> Post(Register request)
        {

            if (!Request.IsInProcessRequest())
            {
                throw HttpError.MethodNotAllowed("Not allowed");
            }
```

The first screenshot show what happens when tests are run in "parallel" ( run all unit tests)
![Capture1](https://user-images.githubusercontent.com/38968/68323198-f55d9e80-0092-11ea-9ef2-3bb917c676f6.JPG)

The second screenshot shows what happens when tests are ran one by one. Note that a single test, even ran alone, can sometimes fail or succeed. This seems quite random.
![Capture2](https://user-images.githubusercontent.com/38968/68323218-00183380-0093-11ea-8962-f0f028c35737.JPG)

The third screenshot show what happens when I remove the `IsInProcessRequest()` check
![Capture3](https://user-images.githubusercontent.com/38968/68323246-0efee600-0093-11ea-9810-ac3a0407af74.JPG)

**Comment by mythz** [comment]
There can only be 1 AppHost configured at 1 time so you never be running integration tests in parallel.

Please provide a stand-alone test I can run that repro's the issue. [Reported Issues always need to be reproducible](https://github.com/ServiceStack/Issues#you-can-view-existing-issues-or-report-a-new-issue) so the underlying cause can be identified that's in an actual issue and not working as designed, e.g. InProcess Service Gateway requests are marked as In Process.

**Comment by ludydoo** [comment]
I managed to create a small project illustrating the error.
The error is intermittent. You have to run the tests a few times to see it happen. 

From my investigation, I believe that the problem arises when a service calls the Gateway.SendAsync more than once for the same request. The problem is intermittent, so it's probably some race condition / multithreading timing thing (perhaps)? Not too familiar with how the GrpcClient works. 

I attached the zip for you.

[grpcerror.zip](https://github.com/ServiceStack/Issues/files/3820283/grpcerror.zip)

**Comment by ludydoo** [comment]
This is an updated list of tests

```
using System.Collections.Generic;
using System.Threading.Tasks;
using grpcerror.ServiceModel.Tec.ServiceModel.Tenants;
using NUnit.Framework;

namespace grpcerror.Tests
{
    public class IntegrationTest : IntegrationTestBase
    {
        private static          int    _count;
        private static readonly object Lock = new object();


        private static SignUpNewTenant GetDto(bool callGatewayTwiceInternally)
        {
            lock (Lock)
            {
                _count++;
                var email       = $"email{_count}@email.com";
                var tenancyName = $"tenancy{_count}";
                var password    = "password";
                var signupTenant = new SignUpNewTenant
                {
                    AdminEmail       = email,
                    AdminPassword    = password,
                    ConfirmPassword  = password,
                    AutoLogin        = true,
                    CompanyName      = "company",
                    TenancyName      = tenancyName,
                    CallGatewayTwice = callGatewayTwiceInternally
                };
                return signupTenant;
            }
        }

        [Test]
        [Repeat(1000)]
        public async Task Calling_Service_Gateway_Once_In_Service_Repeated()
        {
            using var client = CreateClient();
            await client.PostAsync(GetDto(false));
            Assert.Pass();
        }


        [Test]
        public async Task Calling_Service_Gateway_Twice_In_Service()
        {
            using var client = CreateClient();
            await client.PostAsync(GetDto(true));
            Assert.Pass();
        }

        [Test]
        [Repeat(100)]
        public async Task Calling_Service_Gateway_Twice_In_Service_Repeated()
        {
            using var client = CreateClient();
            await client.PostAsync(GetDto(true));
            Assert.Pass();
        }

        [Test]
        [Repeat(10)]
        public async Task Calling_Service_Gateway_Twice_In_Service_Repeated_With_Delay()
        {
            await Task.Delay(1000);
            using var client = CreateClient();
            await client.PostAsync(GetDto(true));
            Assert.Pass();
        }

        [Test]
        public void Calling_Service_Gateway_Twice_In_Service_Parallel()
        {
            var tasks = new List<Task>();
            for (var i = 0; i < 1000; i++)
                tasks.Add(Task.Run(() =>
                {
                    var client = CreateClient();
                    return client.PostAsync(GetDto(true));
                }));
            Task.WaitAll(tasks.ToArray());
        }

        [Test]
        [Repeat(10)]
        public void Calling_Service_Gateway_Twice_In_Service_Parallel_Repeated()
        {
            var tasks = new List<Task>();
            for (var i = 0; i < 1000; i++)
                tasks.Add(Task.Run(() =>
                {
                    var client = CreateClient();
                    return client.PostAsync(GetDto(true));
                }));
            Task.WaitAll(tasks.ToArray());
        }

        [Test]
        [Repeat(10)]
        public async Task Calling_Service_Gateway_Twice_In_Service_Parallel_Repeated_With_Delay()
        {
            await Task.Delay(1000);
            var tasks = new List<Task>();
            for (var i = 0; i < 1000; i++)
                tasks.Add(Task.Run(() =>
                {
                    var client = CreateClient();
                    return client.PostAsync(GetDto(true));
                }));
            Task.WaitAll(tasks.ToArray());
        }
    }
}
```

**Comment by ludydoo** [comment]
Could it be possible that : 

1. Request arrives at TenantService
2. The InProcessServiceGateway sets the request as InProcess
3. The request gets executed by the PartyService
4. The InProcessServiceGateway sets the request as not InProcess
5. The ManagedServiceExec does `HostContext.Async.ContinueWith(request, taskResponse, task => appHost.Release(service));`
6. The TenantService calls InProcessServiceGateway (RegisterService)
7. The InProcessServiceGateway sets the request as InProcess
7. The async task from #5 sets the process as not InProcess
8. It fails. 

I see that the gateway uses the same Request instance, and just changes the dto and the verb to pass it on the pipeline again. But I'm thinking that there is a race condition when the services are disposed. The ManagedServiceExec -> Service.Dispose -> ReleaseIfInProcessRequest happens asynchronously. So the request object might be marked as not InProcess asynchronously. Which would interfere with a successive call to the service gateway. 

There would be a race condition on request.RequestAttributes between the InProcessServiceGateway, and the ManagedServiceExec.Release (which calls Service.Release wich calls ReleaseIfInProcessRequest). 

I believe that to confirm this, we would have to add a delay between the two ServiceGateway.SendAsync calls.

**Comment by ludydoo** [comment]
Updated zip containing tests with delay between the two Gateway.Send calls

[grpcerror.zip](https://github.com/ServiceStack/Issues/files/3821037/grpcerror.zip)

**Comment by mythz** [comment]
Thanks for the diagram, that's definitely over and above the call for a repro :)

I'm able to repro the issue with your example, I'm looking at creating a repro in ServiceStack integration tests.

Basically the in process ServiceGateway should support async but not running concurrently in parallel as `IRequest` instance is not ThreadSafe. So it should be complete before the 2nd Service Gateway request is done. So basically need to look at why that's not happening.

**Comment by mythz** [comment]
This should now be resolved from [this commit](https://github.com/ServiceStack/ServiceStack/commit/efdc433dbb121ace9e787ed3c0b79a5edbcb4c1d).

This change is available from the latest v5.7.1 that's now [available on MyGet](https://docs.servicestack.net/myget#redownloading-myget-packages).

**Comment by ludydoo** [comment]
Hi,
Thanks for the quick fix. 
Though, this does not fully work.

For example, the `TenantService` calls `RegisterService.Register`.
But there is a `AbstractValidator<Register>`
In which there is a call to `Gateway.GetParty(...)`
Which will set the `IsInProcessRequest` to `false`

```
public class TenantService : Service {
  Post(SignUpNewTenant input){
    ... 
    Gateway.SendAsync(new Register())
    ... 
  }
}

public class RegisterValidator : AbstractValidator<Register>{
  void Validate(Register input){
    ... 
    Gateway.SendAsync(new GetParty())
    ...
  }
}

public class RegisterService : Service {
   public object Post(Register){
      if (!Request.IsInProcessRequest()) throw HttpError.MethodNotAllowed();
   }
}

```

I can verify that by removing the call on the `RegisterValidator` to `Gateway.SendAsync` and the tests succeed ( request is in process )

**Comment by mythz** [comment]
Nested Gateway calls should now be resolved from [this commit](https://github.com/ServiceStack/ServiceStack/commit/a8175fa36a4f4eec700d02ac70057c6c2ce25a53) available in the latest v5.7.1 that's now on MyGet.
